### PR TITLE
[State Observability] pre-alpha documentation

### PR DESCRIPTION
**Pull request by rkooo567**
Adds
1. Documentation for state APIs
2. API reference

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/experimental/state/custom_types.py
Comment (rkooo567) [inline python/ray/experimental/state/custom_types.py:17]: Don't really like how I did this, but I am not surer what's the better alternative that's not overengineering (like auto generating this file). Looks like we cannot access the grpc enum when generating a doc because they mock them out not to compile ray before creating a doc (ping me if you'd like to talk this in person)

File: dashboard/memory_utils.py
Comment (rickyyx) [inline dashboard/memory_utils.py:51]:
```suggestion

```

File: python/ray/experimental/state/custom_types.py
Comment (rickyyx) [reply python/ray/experimental/state/custom_types.py:17]: Is the purpose of this so that we could reference the enum values in doc? 
Could we circumvent it by not explicitly listing all the enum values for the associated status in the docstring? But ask users to defer to the docs of the sources (protobuf or when these enums are documented). I feel the datasource should actually be where the definition of these states live, rather than in the state API since IIUC, we are simply mirroring them.

File: python/ray/experimental/state/common.py
Comment (rickyyx) [inline python/ray/experimental/state/common.py:162]: oh no, i have not been typing my fullstops...

File: python/ray/scripts/scripts.py
Comment (scv119) [inline python/ray/scripts/scripts.py:2086]: looks follow is a boolean? also should all arguments `--` instead of `-`?

File: python/ray/experimental/state/common.py
Comment (scv119) [inline python/ray/experimental/state/common.py:600]: hmm what's special of `#:`?

File: python/ray/experimental/state/state_cli.py
Comment (rickyyx) [inline python/ray/experimental/state/state_cli.py:334]: We probably want to embed constant values here or reference the exact consts or remember to change it once we figure out the fixed system upperbound for this.

File: python/ray/experimental/state/state_cli.py
Comment (rickyyx) [inline python/ray/experimental/state/state_cli.py:392]: For CLIs, I think it would be nice to always include examples in the docstring, so when people do `--help`, they could use those directly. Like the `ray logs` one below

File: python/ray/scripts/scripts.py
Comment (rickyyx) [inline python/ray/scripts/scripts.py:2565]: nit: maybe we should also do `ray_list` like what you did here. `ray_xxx` is definitely a better than my `state_cli_xxx` lol

**Review by rickyyx — APPROVED** [review]
Could I see the generated docs hosted somewhere? Or I need to manually build the doc on this PR to see it?

**Comment by rkooo567** [comment]
Screen Shot 2022-07-14 at 3 31 20 PM

@rickyyx you can see it from here! Looks like there's a bug..

**Comment by rkooo567** [comment]
maybe because `Literal` is not available in the lower pytohn version. I will fix this

File: python/ray/experimental/state/state_cli.py
Comment (rkooo567) [reply python/ray/experimental/state/state_cli.py:392]: Ah good point. I removed them because it is auto-generated by click, but this is not available when --help is used..

**Review by scv119 — APPROVED** [review]

**Comment by rkooo567** [comment]
The doc will be also added to this PR. will be ready by tomorrow

**Comment by rkooo567** [comment]
@rickyyx @scv119 I also added a documentation here. Please take a look at https://readthedocs.org/projects/ray/builds/17512917/ and `state-api.rst`

File: python/ray/experimental/state/state_cli.py
Comment (rkooo567) [inline python/ray/experimental/state/state_cli.py:35]: @rickyyx I removed it because we can add this information to the top-level document (and it also screws up the formatting of docstring for some reasons. Tried fixing it, but no luck)

File: doc/source/ray-observability/state/state-api.rst
Comment (scv119) [inline doc/source/ray-observability/state/state-api.rst:289]: Finished is a bit confusing... maybe just name it garbage collected?

File: doc/source/ray-observability/state/state-api.rst
Comment (scv119) [inline doc/source/ray-observability/state/state-api.rst:278]: incorrect -> incomplete

File: doc/source/ray-observability/state/state-api.rst
Comment (scv119) [inline doc/source/ray-observability/state/state-api.rst:284]: size or number of rows? need a bit clarity

File: doc/source/ray-observability/state/state-api.rst
Comment (rickyyx) [inline doc/source/ray-observability/state/state-api.rst:42]: With this small workload, it should return quite quickly. Should we mention it is only expected to be slow when querying large number of entities maybe at a later section? 

```suggestion
Now, let's see the summarized states of tasks. 
```

File: doc/source/ray-observability/state/state-api.rst
Comment (rickyyx) [inline doc/source/ray-observability/state/state-api.rst:79]: image

Not really relate to this PR, more of a cosmetic minor point. Should we change the default columns set for task to exclude "NAME" column? On a narrow terminal or window like this, it is not showing up too well.

File: doc/source/ray-observability/state/state-api.rst
Comment (rickyyx) [inline doc/source/ray-observability/state/state-api.rst:89]:
```suggestion
    ray get tasks 32d950ec0ccf9d2affffffffffffffffffffffff01000000
```

File: doc/source/ray-observability/state/state-api.rst
Comment (rickyyx) [inline doc/source/ray-observability/state/state-api.rst:123]: should we promote this to the beginning of this page?

File: doc/source/ray-observability/state/state-api.rst
Comment (rickyyx) [inline doc/source/ray-observability/state/state-api.rst:125]: mention log as well? for completeness?

File: doc/source/ray-observability/state/state-api.rst
Comment (rickyyx) [reply doc/source/ray-observability/state/state-api.rst:79]: should we use `ray list actors` and `ray get actors xxx` for the list results example?

File: doc/source/ray-observability/state/state-api.rst
Comment (rickyyx) [inline doc/source/ray-observability/state/state-api.rst:163]: image
need new line?

File: doc/source/ray-observability/state/state-api.rst
Comment (rickyyx) [inline doc/source/ray-observability/state/state-api.rst:140]: is there a way we make this a header so that it appears in the content section? What do y'all think? So people get a sense of what the APIs could do w/o reading all of them. 

image

File: doc/source/ray-observability/state/state-api.rst
Comment (rickyyx) [inline doc/source/ray-observability/state/state-api.rst:186]: This is not rendered well. 
image

File: doc/source/ray-observability/state/state-api.rst
Comment (rickyyx) [inline doc/source/ray-observability/state/state-api.rst:270]: Oh right - we need to talk about Python here also I think. Maybe throw in some examples. 

Or maybe we could do this for all the examples (CLI and Python vs SSH and HTTPS) : 
image

File: doc/source/ray-observability/state/state-api.rst
Comment (rickyyx) [inline doc/source/ray-observability/state/state-api.rst:269]:
```suggestion
The state APIs don't guarantee to return a consistent/complete snapshot of the cluster all the time. By default,
```

File: doc/source/ray-observability/state/state-api.rst
Comment (rickyyx) [inline doc/source/ray-observability/state/state-api.rst:285]:
```suggestion
Currently, there's no way to avoid truncated data if the query result is too large . In this case, it is recommended to
```

File: doc/source/ray-observability/state/state-api.rst
Comment (rickyyx) [inline doc/source/ray-observability/state/state-api.rst:286]: This is actually not true right? Since we do truncation before applying filters. I don't think filters will help.

File: doc/source/ray-observability/state/state-api.rst
Comment (rickyyx) [inline doc/source/ray-observability/state/state-api.rst:299]: This looks weird though. Should we just promote this to a top level header?  
image

File: python/ray/experimental/state/custom_types.py
Comment (rickyyx) [reply python/ray/experimental/state/custom_types.py:22]: I see - I agree. After looking at the generated doc, I think documenting these states make sense.

**Review by rickyyx — APPROVED** [review]
This looks really great!



For the filterable and detail options, I wonder if we could embed the `filterable/detail` info into each column's doc, rather than have a separate section? The generated doc looks a bit verbose imo. 
image

File: doc/source/ray-observability/state/state-api.rst
Comment (rkooo567) [reply doc/source/ray-observability/state/state-api.rst:140]: Looks like we don't do this throughout the document. Top level headers only include new document, not headers of each doc. (and I couldn't find a way to convert it..). Will just keep it as it is..

**Review by rkooo567 — COMMENTED** [review]
Addressed all comments. For Python APIs, I will leave it to @rickyyx to update!

Issue tracked: https://github.com/ray-project/ray/issues/26949

**Comment by rkooo567** [comment]
cc @richardliaw can you take a look at this PR? I need approval from a code owner

**Review by edoakes — APPROVED** [review]
codeowner stamp

**Pull request merged**
